File: app/services/variant_units/unit_prices.rb
# frozen_string_literal: true

module VariantUnits
  class UnitPrices
    def initialize(variant)
      @variant = variant
      @product = variant.product
    end

    def denominator
      # catches any case where unit is not kg, lb, or L.
      return @variant.unit_value if @product&.variant_unit == "items"

      case unit
      when "lb"
        @variant.unit_value / 453.6
      when "kg"
        @variant.unit_value / 1000
      else # Liters
        @variant.unit_value
      end
    end

    def unit
      return "lb" if VariantUnits::WeightsAndMeasures.new(@variant).
        system_of_measurement == "imperial"

      case @product&.variant_unit
      when "weight"
        "kg"
      when "volume"
        "L"
      else
        @product.variant_unit_name.presence || "item"
      end
    end
  end
end

File: app/webpacker/js/services/unit_prices.js
import PriceParser from "js/services/price_parser";
import VariantUnitManager from "js/services/variant_unit_manager";
import localizeCurrency from "js/services/localize_currency";

export default class UnitPrices {
  constructor() {
    this.variantUnitManager = new VariantUnitManager();
    this.priceParser = new PriceParser();
  }

  displayableUnitPrice(price, scale, unit_type, unit_value, variant_unit_name) {
    price = this.priceParser.parse(price);
    if (price && !isNaN(price) && unit_type && unit_value) {
      const value = localizeCurrency(
        this.price(price, scale, unit_type, unit_value, variant_unit_name),
      );
      const unit = this.unit(scale, unit_type, variant_unit_name);
      return `${value} / ${unit}`;
    }
    return null;
  }

  price(price, scale, unit_type, unit_value) {
    return price / this.denominator(scale, unit_type, unit_value);
  }

  denominator(scale, unit_type, unit_value) {
    const unit = this.unit(scale, unit_type);
    if (unit === "lb") {
      return unit_value / 453.6;
    } else if (unit === "kg") {
      return unit_value / 1000;
    } else {
      return unit_value;
    }
  }

  unit(scale, unit_type, variant_unit_name = "") {
    if (variant_unit_name.length > 0 && unit_type === "items") {
      return variant_unit_name;
    } else if (unit_type === "items") {
      return "item";
    } else if (this.variantUnitManager.systemOfMeasurement(scale, unit_type) === "imperial") {
      return "lb";
    } else if (unit_type === "weight") {
      return "kg";
    } else if (unit_type === "volume") {
      return "L";
    }
  }
}
